packages/source-map/lib/sourceMap.ts

```diff
@@ -11,76 +11,140 @@ export interface Mapping<Data = unknown> {
+			}
+		}
 
-				const mapped = translateOffset(offset, mapping[fromRange], mapping[toRange], getLengths(mapping, fromRange), getLengths(mapping, toRange));
-				if (mapped !== undefined) {
-					yield [mapped, mapping as Mapping<Data>] as const;
+		const toRange: CodeRangeKey = fromRange == 'sourceOffsets' ? 'generatedOffsets' : 'sourceOffsets'
+		if (perfectMatches.length > 0) {
+			// Prefer the shortest mapping ranges
+			perfectMatches.sort((a, b) => getSpanLength(a, fromRange) - getSpanLength(b, fromRange))
+
+			for (const match of perfectMatches) {
+				const fromStartOffset = getStartOffset(match, fromRange)
+				const toStartOffset = getStartOffset(match, toRange)
+				const toLength = getSpanLength(match, toRange)
+				const fromLength = getSpanLength(match, fromRange)
+				if (toLength == fromLength) {
+					// `from` and `to` span mappings have the same length - map range verbatim
+					const startOffset = toStartOffset + start - fromStartOffset
+					yield [startOffset, startOffset + length, match[0]]
+				} else if (fromStartOffset == start && fromLength == length) {
+					// The whole `from` span is selected - map to the whole `to` span
+					yield [toStartOffset, toStartOffset + toLength, match[0]]
+				} else {
+					// We would need to do some heuristics here to map the span, and it would make little
+					// sense in terms of mapping accuracy.
+					// Try the next match.
+				}
+			}
+		} else {
+			// Prefer the shortest `from` spans
+			startMatches.sort((a, b) =>
+				getSpanLength(a, fromRange) - getSpanLength(b, fromRange))
+			endMatches.sort((a, b) =>
+				getSpanLength(a, fromRange) - getSpanLength(b, fromRange))
+
+			for (const startMatch of startMatches) {
+				let mapping = startMatch[0];
+
+				for (const endMatch of endMatches) {
+					if (endMatch[0] != mapping) {
+						continue
+					}
+					const startOffset = mapOffset(startMatch, start, fromRange, toRange)
+					const endOffset = mapOffset(endMatch, end, fromRange, toRange)
+					yield [startOffset, endOffset, mapping]
 				}
 			}
 		}
 	}
 
-	private getMemoBasedOnRange(fromRange: CodeRangeKey) {
+	private getStorageBasedOnRange(fromRange: CodeRangeKey) {
 		return fromRange === 'sourceOffsets'
-			? this.sourceCodeOffsetsMemo ??= this.createMemo('sourceOffsets')
-			: this.generatedCodeOffsetsMemo ??= this.createMemo('generatedOffsets');
+			? this.sourceCodeOffsetsStorage ??= this.createStorage('sourceOffsets')
+			: this.generatedCodeOffsetsStorage ??= this.createStorage('generatedOffsets');
 	}
 
-	private createMemo(key: CodeRangeKey): MappingMemo {
+	private createStorage(key: CodeRangeKey): BinarySearchStorage<Data> {
+		if (!this.mappings.every(mapping => areRangesSortedAndNonOverlapping(mapping[key], getLengths(mapping, key)))) {
+			throw new Error("Ranges within one mapping should be sorted and non-overlapping.")
+		}
+		return new BinarySearchStorage<Data>(this.mappings, key)
+	}
+
+}
+
+class BinarySearchStorage<Data> {
```

Comment:
I've originally had the interval storage here, which turned out to be inefficient and not really needed anyway, so I kept the logic in the separate class.